File: problems/Battleships in a Board/countBattleships.cs
public class Solution {
    public int CountBattleships(char[][] board) {
        (int dx, int dy)[] _directions = {(0,-1), (-1,0)};

        int n = board.Length;

        if (n == 0)
        {
            return 0;
        }

        int m = board[0].Length;

        int res = 0;


        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (board[i][j] == 'X')
                {
                    bool adj = false;

                    foreach (var dir in _directions)
                    {
                        int newI = i + dir.dx;
                        int newJ = j + dir.dy;

                        if (newI >= 0 && newI < n && newJ >= 0 && newJ < m && board[newI][newJ] == 'X')
                        {
                            adj = true;
                            break;
                        }
                    }

                    if (!adj)
                    {
                        res++;
                    }
                }
            }
        }

        return res;
    }
}

File: problems/Battleships in a Board/countBattleships.java
class Solution {
    public int countBattleships(char[][] board) {
        if (0 == board.length) {
            return 0;
        }

        int[][] directions = new int[][] {{-1,0},{0,-1}};
        int rows = board.length;
        int cols = board[0].length;
        int result = 0;

        for (int row = 0; rows > row; ++row) {
            for (int col = 0; cols > col; ++col) {
                if ('X' == board[row][col]) {
                    boolean flag = false;

                    for (int[] dir : directions) {
                        int newRow = row + dir[0];
                        int newCol = col + dir[1];

                        if (0 <= newRow && rows > newRow && 0 <= newCol && cols > newCol && 'X' == board[newRow][newCol]) {
                            flag = true;
                            break;
                        }
                    }

                    if (!flag) {
                        ++result;
                    }
                }
            }
        }

        return result;
    }
}
